clients/src/main/java/org/apache/kafka/common/requests/DescribeClusterRequest.java

```diff
@@ -51,6 +52,13 @@ public String toString() {
     public DescribeClusterRequest(DescribeClusterRequestData data, short version) {
         super(ApiKeys.DESCRIBE_CLUSTER, version);
         this.data = data;
+        validateVersionForIncludingFencedBrokers(version, data.includeFencedBrokers());
+    }
+
+    private void validateVersionForIncludingFencedBrokers(short version, boolean includeFencedBrokers) {
+        if (version < 2 && includeFencedBrokers) {
+            throw new UnsupportedVersionException("Including fenced broker endpoints is not supported with version " + version);
+        }
```

Comment:
In the generated clients/build/generated/main/java/org/apache/kafka/common/message/DescribeClusterRequestData.java, the `UnsupportedVersionException` will be thrown when trying to add `includeFencedBrokers` into version < 2, so we don't have to do it by our own here.

Reply:
This doesn't seem to be the case when I tested it by running new AdminClient against an old broker. For example, if we remove this validation, testDescribeClusterHandleUnsupportedVersionForIncludingFencedBrokers() test fails. Also this helps Admin to handle the UnsupportedVersion exception differently, otherwise we fall back to metadata request and it's not clear to users why the request the failed.

Reply:
1. Currently, the `testDescribeClusterHandleUnsupportedVersionForIncludingFencedBrokers` will return `UnsupportedVersionException` with `Api DESCRIBE_CLUSTER with version 1` from [here](https://github.com/apache/kafka/blob/48d60efe9802786451458aec9136b7b585fecd92/clients/src/test/java/org/apache/kafka/clients/MockClient.java#L261). That's why it will fail. So, basically, it will fail with `UnsupportedVersionException` as what we expected. 

2. I understand why we need this different message as described in this [comment](https://github.com/apache/kafka/pull/17524#discussion_r1850837344).
> We fallback to metadata request, if the broker does not support the new DescribeCluster API. In this case, DescribeCluster API is supported but the option to include fenced broker is not. So this is why we are checking the specific error message to differentiate the cause of UnsupportedVersion exception.

But I think using the error message as a condition is really weak. I think, like you said, the reason we can't fallback to metadata request is because the user added an option to _include fenced broker_. That is, when user requests to "include fenced broker", this is the case that we should not fallback to metadata request, even if it returned `UnsupportedVersionException`. After all, the metadata request response won't include fenced broker, it is meaningless to users. So, like how we treat `usingBootstrapControllers`, we can do similar thing to this case, WDYT?

That is:
```
boolean handleUnsupportedVersionException(final UnsupportedVersionException exception) {
  if (metadataManager.usingBootstrapControllers()) {
     return false;
  }
  // add some comments here
  if (options.includeFencedBrokers()) {
    return false;
  }
```

Also, we should have an integration test for this unsupported version case. Like [this test](https://github.com/apache/kafka/blob/48d60efe9802786451458aec9136b7b585fecd92/core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala#L1354) did.

Reply:
I agree that checking the option instead of the error message is much better solution here. I made an update to add an error level log, instead of validating describe request and throwing a specific error for includeFencedBrokers option. 

In terms of the integration test, I'm going to look into to it a little more. KRaftClusterTest.scala you pointed, seems like more cluster level tests. Since this is testing Admin specific option, it should probably be added to AdminIntegrationTest but I'm currently not sure how to set the test cluster's version or it's supported API version.

Reply:
Hmm... it's quite complicated to have a integration test to test with lower describeCluster version. I think it's fine sine we already have unit test to cover it. Thanks.

Reply:
I agree, I think this will take quite a lot of refactoring to the testing framework. We have unit tests for the compatibility part and integration tests for the new option to include fenced brokers.